x/uniswap/handler.go

```diff
@@ -0,0 +1,152 @@
+	}
+
+	nativeLiqudiity, err := k.GetExchange(ctx, NativeAsset)
+	if err != nil {
+		panic("error retrieving native asset total liquidity")
+	}
+
+	totalUNI, err := k.GetTotalUNI(ctx)
+	if err != nil {
+		panic("error retrieving total UNI")
+	}
+
+	// calculate amount of UNI to be minted for sender
+	// and coin amount to be deposited
+	MintedUNI := (totalUNI.Mul(msg.DepositAmount)).Quo(nativeLiquidity)
+	coinDeposited := (totalUNI.Mul(msg.DepositAmount)).Quo(nativeLiquidity)
+	nativeCoin := sdk.NewCoin(NativeAsset, msg.DepositAmount)
+	exchangeCoin := sdk.NewCoin(msg.ExchangeDenom, coinDeposited)
+
+	// transfer deposited liquidity into uniswaps ModuleAccount
+	err := k.SendCoinsAccountToModule(ctx, msg.Sender, ModuleName, sdk.NewCoins(nativeCoin, coinDeposited))
+	if err != nil {
+		return err
+	}
+
+	// set updated total UNI
+	totalUNI = totalUNI.Add(MintedUNI)
+	k.SetTotalUNI(ctx, totalUNI)
+
+	// update senders account with minted UNI
+	UNIBalance := k.GetUNIForAddress(ctx, msg.Sender)
+	UNIBalance = UNIBalance.Add(MintedUNI)
+	k.SetUNIForAddress(ctx, UNIBalance)
+
+	return sdk.Result{}
+}
+
+// HandleMsgRemoveLiquidity handler for MsgRemoveLiquidity
+func HandleMsgRemoveLiquidity(ctx sdk.Context, msg MsgRemoveLiquidity, k Keeper) sdk.Result {
+	// check if exchange exists
+	coinLiquidity, err := k.GetExchange(ctx, msg.ExchangeDenom)
+	if err != nil {
+		panic(fmt.Sprintf("error retrieving total liquidity for denomination: %s", msg.ExchangeDenom))
+	}
+
+	nativeLiquidity, err := k.GetExchange(ctx, NativeAsset)
+	if err != nil {
+		panic("error retrieving native asset total liquidity")
+	}
+
+	totalUNI, err := k.GetTotalUNI(ctx)
+	if err != nil {
+		panic("error retrieving total UNI")
+	}
+
+	// calculate amount of UNI to be burned for sender
+	// and coin amount to be returned
+	nativeWithdrawn := msg.WithdrawAmount.Mul(nativeLiquidity).Quo(totalUNI)
+	coinWithdrawn := msg.WithdrawAmount.Mul(coinLiqudity).Quo(totalUNI)
+	nativeCoin := sdk.NewCoin(nativeDenom, nativeWithdrawn)
+	exchangeCoin = sdk.NewCoin(msg.ExchangeDenom, coinWithdrawn)
+
+	// transfer withdrawn liquidity from uniswaps ModuleAccount to sender's account
+	err := k.SendCoinsModuleToAccount(ctx, msg.Sender, ModuleName, sdk.NewCoins(nativeCoin, coinDeposited))
+	if err != nil {
+		return err
+	}
+
+	// set updated total UNI
+	totalUNI = totalUNI.Add(MintedUNI)
+	k.SetTotalUNI(ctx, totalUNI)
+
+	// update senders account with minted UNI
+	UNIBalance := k.GetUNIForAddress(ctx, msg.Sender)
+	UNIBalance = UNIBalance.Add(MintedUNI)
+	k.SetUNIForAddress(ctx, UNIBalance)
+
+	return sdk.Result{}
+}
+
+// GetInputAmount returns the amount of coins sold (calculated) given the output amount being bought (exact)
+// The fee is included in the output coins being bought
+// https://github.com/runtimeverification/verified-smart-contracts/blob/uniswap/uniswap/x-y-k.pdfhttps://github.com/runtimeverification/verified-smart-contracts/blob/uniswap/uniswap/x-y-k.pdf
```

Comment:
all these links should go on a spec with references instead. Can be addressed on a separate PR though